### PR TITLE
Implement GraphQL Security BBEs

### PR DESCRIPTION
**Pull request by ldclakmal**
## Purpose
$subject was done for following BBEs:

**GraphQL Security**
- Service - Basic Auth File User Store
- Service - Basic Auth LDAP User Store
- Service - JWT Auth
- Service - OAuth2
- Service - SSL/TLS
- Service - Mutual SSL

Related to https://github.com/ballerina-platform/ballerina-standard-library/issues/584

### CODE REVIEW

File: new-examples/graphql-service-basic-auth-file-user-store/graphql_service_basic_auth_file_user_store.description [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// An GraphQL service can be secured using Basic auth and optionally by
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-basic-auth-file-user-store/graphql_service_basic_auth_file_user_store.description:1]:
```suggestion
// A GraphQL service can be secured using Basic auth and optionally by
```

File: new-examples/graphql-service-basic-auth-file-user-store/graphql_service_basic_auth_file_user_store.description [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+// enforcing authorization. Then, it validates the Basic auth token sent in the
+// `Authorization` header against the provided configurations. This reads data
+// from a file, which has a TOML format. This stores the usernames, passwords
+// for authentication, and scopes for authorization.<br/>
+// Ballerina uses the concept of scopes for authorization. A resource declared
+// in a service can be bound to one/more scope(s).<br/>
+// In the authorization phase, the scopes of the service/resource are compared
+// against the scope included in the user store for at least one match between
+// the two sets.<br/>
+// `Config.toml` has defined three users - Alice, Bob and Eve. Each user has a
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-basic-auth-file-user-store/graphql_service_basic_auth_file_user_store.description:11]: These are different from the actual values, right? Shall we update them?

File: new-examples/graphql-service-basic-auth-ldap-user-store/graphql_service_basic_auth_ldap_user_store.bal [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        }
+    }
+);
+
+// The service can be secured with Basic auth and can be authorized  optionally.
+// Basic auth using the LDAP user store can be enabled by setting the
+// `graphql:LdapUserStoreConfig` configurations.
+// Authorization is based on scopes. A scope maps to one or more groups.
+// Authorization can be enabled by setting the `string|string[]` type
+// configurations.
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-basic-auth-ldap-user-store/graphql_service_basic_auth_ldap_user_store.bal:17]: This is a little bit unclear (for me).

File: new-examples/graphql-service-basic-auth-ldap-user-store/graphql_service_basic_auth_ldap_user_store.description [hunk 4]
```diff
@@ -0,0 +1,12 @@
+// An GraphQL service can be secured using Basic auth and by enforcing
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-basic-auth-ldap-user-store/graphql_service_basic_auth_ldap_user_store.description:1]:
```suggestion
// A GraphQL service can be secured using Basic auth and by enforcing
```

File: new-examples/graphql-service-jwt-auth/graphql_service_jwt_auth.bal [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        }
+    }
+);
+
+// The service can be secured with JWT authentication and can be authorized
+// optionally. JWT authentication can be enabled by setting the
+// `graphql:JwtValidatorConfig` configurations.
+// Authorization is based on scopes. A scope maps to one or more groups.
+// Authorization can be enabled by setting the `string|string[]` type
+// configurations.
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-jwt-auth/graphql_service_jwt_auth.bal:17]: Same as above

File: new-examples/graphql-service-jwt-auth/graphql_service_jwt_auth.description [hunk 6]
```diff
@@ -0,0 +1,12 @@
+// An GraphQL service can be secured using JWT and by enforcing
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-jwt-auth/graphql_service_jwt_auth.description:1]:
```suggestion
// A GraphQL service can be secured using JWT and by enforcing
```

File: new-examples/graphql-service-jwt-auth/graphql_service_jwt_auth.description [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+// authorization optionally. Then, it validates the JWT sent in the
+// `Authorization` header against the provided configurations.<br/>
+// Ballerina uses the concept of scopes for authorization. A resource declared
+// in a service can be bound to one/more scope(s). The scope can be included
+// in the JWT using a custom claim attribute. That custom claim attribute
+// also can be configured as the `scopeKey`.<br/>
+// In the authorization phase, the scopes of the service/resource are compared
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-jwt-auth/graphql_service_jwt_auth.description:8]: Can this be applied to the resource level?

File: new-examples/graphql-service-mutual-ssl/graphql_service_mutual_ssl.bal [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import ballerina/graphql;
+import ballerina/http;
+
+// An GraphQL listener can be configured to accept new connections that are
+// secured via mutual SSL.
+// The [graphql:ListenerSecureSocket](https://docs.central.ballerina.io/ballerina/graphql/latest/records/ListenerSecureSocket) record provides the SSL-related listener configurations.
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-mutual-ssl/graphql_service_mutual_ssl.bal:6]:
```suggestion
// The [`graphql:ListenerSecureSocket`](https://docs.central.ballerina.io/ballerina/graphql/latest/records/ListenerSecureSocket) record provides the SSL-related listener configurations.
```

File: new-examples/graphql-service-mutual-ssl/graphql_service_mutual_ssl.bal [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+            verifyClient: http:REQUIRE,
+            cert: "../resource/path/to/public.crt"
+        },
+        // Enables the preferred SSL protocol and its versions.
+        protocol: {
+            name: http:TLS,
+            versions: ["TLSv1.2", "TLSv1.1"]
+        },
+        // Configures the preferred ciphers.
+        ciphers: ["TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"]
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-mutual-ssl/graphql_service_mutual_ssl.bal:24]: To separate the description from the next line.

```suggestion
        ciphers: ["TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"]

```

File: new-examples/graphql-service-oauth2/graphql_service_oauth2.bal [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        }
+    }
+);
+
+// The service can be secured with OAuth2 authentication and can be authorized
+// optionally. OAuth2 authentication can be enabled by setting the
+// `graphql:OAuth2IntrospectionConfig` configurations.
+// Authorization is based on scopes. A scope maps to one or more groups.
+// Authorization can be enabled by setting the `string|string[]` type
+// configurations.
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-oauth2/graphql_service_oauth2.bal:17]: Same comment as above.

File: new-examples/graphql-service-oauth2/graphql_service_oauth2.bal [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+            },
+            scopes: ["admin"]
+        }
+    ]
+}
+service /graphql on securedEP {
+
+    // It is optional to override the authentication and authorization
+    // configurations at the resource levels. Otherwise, the service auth
+    // configurations will be applied automatically to the resources as well.
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-oauth2/graphql_service_oauth2.bal:40]: Is this supported in GraphQL?

File: new-examples/graphql-service-oauth2/graphql_service_oauth2.description [hunk 12]
```diff
@@ -0,0 +1,13 @@
+// An GraphQL service can be secured using OAuth2 and by enforcing
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-oauth2/graphql_service_oauth2.description:1]:
```suggestion
// A GraphQL service can be secured using OAuth2 and by enforcing
```

File: new-examples/graphql-service-ssl-tls/graphql_service_ssl_tls.bal [hunk 13]
```diff
@@ -0,0 +1,21 @@
+import ballerina/graphql;
+
+// An GraphQL listener can be configured to communicate through HTTPS as well.
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-ssl-tls/graphql_service_ssl_tls.bal:3]:
```suggestion
// A GraphQL listener can be configured to communicate through HTTPS as well.
```

File: new-examples/graphql-service-ssl-tls/graphql_service_ssl_tls.bal [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+// To secure an listener using HTTPS, the listener needs to be configured with
+// a certificate file and a private key file for the listener.
+// The [graphql:ListenerSecureSocket](https://docs.central.ballerina.io/ballerina/graphql/latest/records/ListenerSecureSocket) record
```
Comment (ThisaruGuruge) [inline new-examples/graphql-service-ssl-tls/graphql_service_ssl_tls.bal:6]:
```suggestion
// The [`graphql:ListenerSecureSocket`](https://docs.central.ballerina.io/ballerina/graphql/latest/records/ListenerSecureSocket) record
```

**Comment by niveathika** [comment]
@ldclakmal Please sync the branch

File: new-examples/graphql-service-oauth2/graphql_service_oauth2.bal (hunk 11 above)
Comment (ldclakmal) [reply new-examples/graphql-service-oauth2/graphql_service_oauth2.bal]: Fixed with https://github.com/ballerina-platform/ballerina-distribution/pull/1816/commits/60e9f102c550acf74377bc2d648196efd76da06d

File: new-examples/graphql-service-basic-auth-file-user-store/graphql_service_basic_auth_file_user_store.description (hunk 2 above)
Comment (ldclakmal) [reply new-examples/graphql-service-basic-auth-file-user-store/graphql_service_basic_auth_file_user_store.description]: Fixed with https://github.com/ballerina-platform/ballerina-distribution/pull/1816/commits/b30eee9eebf69efacbc40cf75017cf4bf36c2695

**Review by praneesha — APPROVED** [review]

**Pull request merged**